### ISSUE TITLE
Avoid replacing class fields with objects of different types

### ISSUE THREAD

**Issue by greenatatlassian** [cekit/cekit#824]
There are a few places where we replace instance variables with things of different type. So the type of something depends on when you access it. IMHO this is a no-no and very prone to errors (possibly explains what is going on with #823).

The most consistent/obvious example is the `Descriptor` (and descendants). The design here is a little strange in that descriptors both provide a way to read YAML descriptor files from disk, including doing schema validation, and a way to represent descriptors for internal purposes. The result is that e.g. what is read from the file as a dictionary appears first in a field as a dictionary, but then is replaced with an object of another type, like `Module`, but that happens to also behave as a dictionary. This replacement is not always perfect (e.g. the dictionary behaviour of `Module` is not guaranteed to match the original dictionary), and it makes the code hard to reason about. 

There are probably several ways we could go about fixing this. My recommendation would be to separate the internal representation from the on disk representation by making them two class hierarchy. We would then have a `YamlDescriptor` and subclasses that represent the exact format expected in the yaml files, including schema validation etc. Then we would have some kind of adaptor that maps e.g. a `YamlModuleDescriptor` into a `Module`. This adaptor would take care of, e.g. standardising types (such as suggested in #822) and could also be the place that any defaults are applied (so we can potentially differentiate whether a user provided a value or we are using a default).  The shape of `Module` then need bear no resemblance to the shape of `YamlModuleDescriptor` (though there will probably be similarities). This will free the code to implement the interface however is best without the interface necessarily being tied to the implementation.

I'm sure there are other ways we could approach this as well. This will be a bit of work, though it need not happen all at once—we can transform one part of the descriptor tree at a time, and in the first instance we would really only be separating the types, not actually making any changes to the shape of either. However, since it is a bit of work, it would be good to make sure we have alignment across the maintainers.

**Comment by greenatatlassian** [comment]
I should add, as time allows, I'm happy to put together a small PR showing how this might look for a single example. Probably would be useful to help gain alignment, unless we decide on a different direction.

**Comment by rnc** [comment]
I agree and I think this is an artifact of the age of the codebase which evolved over time. 

I like the idea of separating out the disk based representation from the in-memory model. I agree that this could be a help for solving #823 - @goldmann - thoughts?

**Comment by goldmann** [comment]
I think it definitely makes sense to improve the current situation!

What I don't like in the current design is that it is not clear what attributes the particular class has. Instead of being clear there we pass it to the underlying dict representation. This makes things hard to read and potentially error-prone as well. Of course at the time when it was implemented (yep, long time ago) it sounded like a great idea! And of course the code back then was simpler and we did not have so many features backed in.

Either case, I'm definitely approving changes in this area! Ideally, I would like to see similar mechanism as in Java with Jackson, which allows for annotations of classes to enable easy (de)searlization. This way we would not really bother with the "internal" representation, because this would be equal to the class representations.

Side note is that we could probably look at replacement of pykwalify as well (no new releases since 2+ years) so that it would suit better in the new world. There are things like [pydantic](https://docs.python-cerberus.org/en/stable/) available. Hmm, https://pydantic-docs.helpmanual.io/usage/exporting_models/#custom-json-deserialisation 😎

**Comment by rnc** [comment]
pydantic looks cool! This could be a good approach.
